### DISCUSSION TITLE
Is it better to use `useTransition` or `useFormStatus` for managing form submission in this scenario or should i go for somethig else?

### DISCUSSION THREAD

**Discussion by theishu4** [vercel/next.js#72396] [Help] [answered]
### Summary

I'm working on a sign-in form in Next, and I'm currently using `useTransition` to manage the form's pending state and updating the other state. But I don't know whether it is good to use `useTransition` hook in this case or not. I want to know if `useTransition` is the best approach here or if there's a better option for handling form state.

Here’s the simplified version of my code:

```javascript
const SigninForm = () => {
  const [isPending, startTransition] = useTransition();
  const [formReturnMsg, setFormReturnMsg] = useState(null);

  const onSubmit = (values) => {
    return startTransition(() => {
     // signinAction is a 'server function'
      signinAction(values).then((value) => {
        setFormReturnMsg(value);
      });
    });
  };

  return (
    <form onSubmit={form.handleSubmit(onSubmit)}>
      <input type="email" disabled={isPending} placeholder="Email" />
      <input type="password" disabled={isPending} placeholder="Password" />
      <button type="submit" disabled={isPending}>
        {isPending ? "Signing in..." : "Sign in"}
      </button>
      {formReturnMsg && <p>{formReturnMsg.message}</p>}
    </form>
  );
};
```

### Questions:

1. Is `useTransition` the most suitable way to handle form submission states like pending and completion in this case?
2. Would `useFormStatus` provide a more efficient or maintainable solution here?
3. Are there any alternative patterns or hooks for handling form submission states effectively in React?

**Comment by icyJoseph** [comment]
I'd say that, https://react.dev/reference/react/useActionState, is what you probably want here, but that uses an `action` prop on the form, not the onSubmit handler.

Mostly because it does, the wiring your are doing here. I think this talk, https://www.youtube.com/watch?v=CptfI2zoleY does a good job at showing, how one progresses from hand made state, to useTransition, to useActionState.

**Reply by theishu4** [reply]
I considered using the action attribute on the form, but I'm encountering an issue with **validation**. When I pass the form data to the action prop, React Hook Form skips client-side validation and sends the data directly to the server function. How can I ensure client-side validation still works in this setup?

I am using the `react-hook-form` with `zod` as resolver.

**Reply by icyJoseph** [reply]
Aha, you left that out initially.

Take a look through here: https://github.com/react-hook-form/react-hook-form/issues/10391#issuecomment-1888225849

**Reply by theishu4** [reply]
so, what should I prefer `useTransition` or creating a new hook? I hope they will support this in future...

**Reply by icyJoseph** [reply, marked as answer]
I think what you are trying to ask, is more on the lane of, using the right tool for the right job.

You should always validate input data on the server. Client side validation is but an accessibility, convenience if you will, for your users, and to ease some load on your server, albeit, minimal. 

This is because, most attackers, would just bypass your client, and post data to your server directly. 

- https://react.dev/reference/rsc/use-server#security
> Arguments to Server Actions are fully client-controlled. For security, always treat them as untrusted input, and make sure to validate and escape arguments as appropriate.

Most form handling libraries, use the `onSubmit` event, rather than `action`, which is what the new React API uses, to consume formActions (server actions). This produces some sort of incompatible layer.  

I think, one has to first recognize that these are React features, server actions and these hooks, and that means, libraries have to offer an API for them. 

Read through here, https://github.com/orgs/react-hook-form/discussions/11832 perhaps, but in general react-hook-forms has to implement their way to work with actions.
